File: apps/rust-api/src/handlers/jobs.rs
//! Job queue API handlers

use axum::{
    extract::{Path, State},
    Json,
};
use serde::{Deserialize, Serialize};
use std::sync::Arc;

use crate::error::{ApiError, ApiResult};
use crate::AppState;
use crate::solver::job_queue::{JobPriority, JobType, JobQueue, QueueStatus};

/// Request to submit a new analysis job
#[derive(Debug, Deserialize)]
pub struct SubmitJobRequest {
    pub job_type: String,       // "static", "modal", "pdelta", "seismic", "buckling", "batch"
    pub priority: Option<String>, // "low", "normal", "high", "urgent"
    pub user_id: String,
    pub input: serde_json::Value,
}

/// Response with job ID
#[derive(Debug, Serialize)]
pub struct SubmitJobResponse {
    pub success: bool,
    pub job_id: String,
    pub message: String,
}

/// Response with job details
#[derive(Debug, Serialize)]
pub struct JobStatusResponse {
    pub success: bool,
    pub job_id: String,
    pub status: String,
    pub progress: f64,
    pub message: String,
    pub created_at: String,
    pub started_at: Option<String>,
    pub completed_at: Option<String>,
    pub result: Option<serde_json::Value>,
    pub error: Option<String>,
}

/// Queue status response
#[derive(Debug, Serialize)]
pub struct QueueStatusResponse {
    pub success: bool,
    pub queue: QueueStatus,
}

fn parse_job_type(s: &str) -> Result<JobType, String> {
    match s.to_lowercase().as_str() {
        "static" => Ok(JobType::StaticAnalysis),
        "modal" => Ok(JobType::ModalAnalysis),
        "pdelta" | "p-delta" => Ok(JobType::PDeltaAnalysis),
        "seismic" => Ok(JobType::SeismicAnalysis),
        "time-history" | "time_history" => Ok(JobType::TimeHistoryAnalysis),
        "buckling" => Ok(JobType::BucklingAnalysis),
        "design" => Ok(JobType::DesignCheck),
        "batch" => Ok(JobType::BatchAnalysis { model_count: 1 }),
        _ => Err(format!("Unknown job type: {}", s)),
    }
}

fn parse_priority(s: Option<&str>) -> JobPriority {
    match s.map(|s| s.to_lowercase()).as_deref() {
        Some("low") => JobPriority::Low,
        Some("high") => JobPriority::High,
        Some("urgent") => JobPriority::Urgent,
        _ => JobPriority::Normal,
    }
}

/// POST /api/jobs - Submit a new job
pub async fn submit_job(
    State(_state): State<Arc<AppState>>,
    Json(req): Json<SubmitJobRequest>,
) -> ApiResult<Json<SubmitJobResponse>> {
    let job_type = parse_job_type(&req.job_type)
        .map_err(|e| ApiError::BadRequest(e))?;
    let priority = parse_priority(req.priority.as_deref());

    // Create a temporary job queue (in production, this would be shared state)
    let (queue, _rx) = JobQueue::new(
        num_cpus::get().max(2)
    );

    let job_id = queue.submit(job_type, priority, req.user_id, req.input).await;

    Ok(Json(SubmitJobResponse {
        success: true,
        job_id: job_id.clone(),
        message: format!("Job {} queued successfully", job_id),
    }))
}

/// GET /api/jobs/:id - Get job status
pub async fn get_job_status(
    State(_state): State<Arc<AppState>>,
    Path(job_id): Path<String>,
) -> ApiResult<Json<JobStatusResponse>> {
    // In production, look up from shared state
    Ok(Json(JobStatusResponse {
        success: true,
        job_id: job_id.clone(),
        status: "queued".into(),
        progress: 0.0,
        message: "Job is in queue".into(),
        created_at: chrono::Utc::now().to_rfc3339(),
        started_at: None,
        completed_at: None,
        result: None,
        error: None,
    }))
}

/// DELETE /api/jobs/:id - Cancel a job
pub async fn cancel_job(
    State(_state): State<Arc<AppState>>,
    Path(job_id): Path<String>,
) -> ApiResult<Json<serde_json::Value>> {
    Ok(Json(serde_json::json!({
        "success": true,
        "job_id": job_id,
        "message": "Job cancellation requested"
    })))
}

/// GET /api/jobs/queue/status - Get queue status
pub async fn get_queue_status(
    State(_state): State<Arc<AppState>>,
) -> ApiResult<Json<QueueStatusResponse>> {
    Ok(Json(QueueStatusResponse {
        success: true,
        queue: QueueStatus {
            queued: 0,
            running: 0,
            completed: 0,
            failed: 0,
            total: 0,
        },
    }))
}

File: apps/backend-python/routers/jobs.py
"""
Job Queue API Endpoints
"""

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel
from typing import Optional, Dict

router = APIRouter(prefix="/api/jobs", tags=["Jobs"])


class JobSubmitRequest(BaseModel):
    job_type: str  # "static", "modal", "pdelta", "buckling", "spectrum"
    priority: Optional[str] = "normal"
    user_id: Optional[str] = None
    input: Dict


@router.post("/submit")
async def submit_analysis_job(req: JobSubmitRequest):
    """Submit a long-running analysis job to the worker pool"""
    try:
        from analysis.worker_pool import get_worker_pool, JobPriority
        pool = await get_worker_pool()
        priority_map = {
            "urgent": JobPriority.URGENT,
            "high": JobPriority.HIGH,
            "normal": JobPriority.NORMAL,
            "low": JobPriority.LOW,
            "batch": JobPriority.BATCH,
        }
        priority = priority_map.get(req.priority or "normal", JobPriority.NORMAL)
        job_id = await pool.submit(req.job_type, req.input, priority, req.user_id)
        return {"success": True, "job_id": job_id, "message": f"Job queued as {req.priority}"}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/{job_id}")
async def get_job_status(job_id: str):
    """Get status of an analysis job"""
    try:
        from analysis.worker_pool import get_worker_pool
        pool = await get_worker_pool()
        job = pool.get_job(job_id)
        if not job:
            raise HTTPException(status_code=404, detail="Job not found")
        return {
            "job_id": job.id,
            "status": job.status.value,
            "job_type": job.job_type,
            "progress": {
                "percent": job.progress.percent,
                "stage": job.progress.stage,
                "message": job.progress.message,
            },
            "result": job.result,
            "error": job.error,
            "created_at": job.created_at,
            "started_at": job.started_at,
            "completed_at": job.completed_at,
        }
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.delete("/{job_id}")
async def cancel_analysis_job(job_id: str):
    """Cancel a queued analysis job"""
    try:
        from analysis.worker_pool import get_worker_pool
        pool = await get_worker_pool()
        cancelled = pool.cancel_job(job_id)
        return {"success": cancelled, "job_id": job_id}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/queue/status")
async def get_queue_status():
    """Get worker queue statistics"""
    try:
        from analysis.worker_pool import get_worker_pool
        pool = await get_worker_pool()
        return pool.get_queue_status()
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))
